core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java

```diff
@@ -4476,6 +4476,85 @@ public String getDefaultScheme() {
+    channelBuilder.maxTraceEvents(10);
+    try {
+      FakeNameResolverFactory nameResolverFactory = new FakeNameResolverFactory
+              .Builder(expectedUri).setServers(Collections.singletonList(new
+              EquivalentAddressGroup(socketAddress))).build();
+      channelBuilder.nameResolverFactory(nameResolverFactory);
+      Map<String, Object> defaultServiceConfig =
+              parseConfig("{\"methodConfig\":[{"
+                      + "\"name\":[{\"service\":\"SimpleService1\"}],"
+                      + "\"waitForReady\":true}]}");
+      channelBuilder.defaultServiceConfig(defaultServiceConfig);
+      createChannel();
+      FakeNameResolverFactory.FakeNameResolver resolver = nameResolverFactory.resolvers.get(0);
+      Status resolutionError = Status.UNAVAILABLE
+              .withDescription("Initial Name Resolution error, using default service config");
+
+      int prevSize = getStats(channel).channelTrace.events.size();
+      resolver.listener.onError(resolutionError);
+      verify(mockLoadBalancer).handleNameResolutionError(resolutionError);
+
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize + 2);
+      assertThat(getStats(channel).channelTrace.events).contains(new ChannelTrace.Event.Builder()
+              .setDescription("Failed to resolve name: " + resolutionError)
+              .setSeverity(ChannelTrace.Event.Severity.CT_WARNING)
+              .setTimestampNanos(timer.getTicker().read())
+              .build());
+
+      prevSize = getStats(channel).channelTrace.events.size();
+      resolver.listener.onError(resolutionError);
+
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize + 1);
+      assertThat(getStats(channel).channelTrace.events).contains(new ChannelTrace.Event.Builder()
+              .setDescription("Failed to resolve name: " + resolutionError)
+              .setSeverity(ChannelTrace.Event.Severity.CT_WARNING)
+              .setTimestampNanos(timer.getTicker().read())
+              .build());
+    } finally {
+      LoadBalancerRegistry.getDefaultRegistry().deregister(mockLoadBalancerProvider);
+    }
+  }
+
+  @Test
+  public void testTraceEventsOnResultsCallFollowedByOnErrorCallWithAppliedServiceConfig()
+          throws Exception {
+    timer.forwardNanos(1234);
+    channelBuilder.maxTraceEvents(10);
+    List<EquivalentAddressGroup> servers = new ArrayList<>();
+    servers.add(new EquivalentAddressGroup(socketAddress));
+    FakeNameResolverFactory nameResolverFactory =
+            new FakeNameResolverFactory.Builder(expectedUri).setServers(servers).build();
+    channelBuilder.nameResolverFactory(nameResolverFactory);
+    createChannel();
+    Status resolutionError = Status.UNAVAILABLE
+            .withDescription("Initial Name Resolution error, using default service config");
+
+    int prevSize = getStats(channel).channelTrace.events.size();
+    ResolutionResult resolutionResult = ResolutionResult.newBuilder()
```

Comment:
Since there is no service config passed in this resolution result, to help the reader understand easier, add a comment above this line that empty service config will get used.

Reply:
Fixed